### PR TITLE
[APT-6294] Android 12 Support

### PR DESCRIPTION
**Pull request by SeanoAndroid**
Updated Gradle, Kotlin version, leak canary, java version, and added mutability flags to pending intents along with new backup format

Release and post-release PRs coming soon

### CODE REVIEW

File: Armadillo/src/main/java/com/scribd/armadillo/Util.kt [hunk 1]
```diff
@@ -43,4 +44,6 @@ fun sanitizeChapters(chapters: List<Chapter>): List<Chapter> {
             }
         }
     }
-}
\ No newline at end of file
+}
+
+fun hasSnowCone() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
```
Comment (SeanoScribd) [inline Armadillo/src/main/java/com/scribd/armadillo/Util.kt:49]: This kind of cascaded into a lot of updated gradle, java, and kotlin versions

File: TestApp/src/main/java/com/scribd/armadillotestapp/presentation/AudioPlayerApplication.kt [hunk 2]
```diff
@@ -34,10 +33,6 @@ class AudioPlayerApplication : Application() {
     override fun onCreate() {
         super.onCreate()
 
-        if (LeakCanary.isInAnalyzerProcess(this)) {
```
Comment (SeanoScribd) [inline TestApp/src/main/java/com/scribd/armadillotestapp/presentation/AudioPlayerApplication.kt:37]: New leak canary has zero boiler plate besides the dependency

File: Armadillo/build.gradle [hunk 3]
```diff
@@ -17,20 +17,21 @@ plugins {
...
 
     defaultConfig {
         minSdkVersion 21
-        targetSdkVersion 30
+        targetSdkVersion 32
         versionCode 1
         versionName project.LIBRARY_VERSION
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
         consumerProguardFiles 'proguard-rules.pro'
+        buildConfigField("String", "VERSION_NAME", "\"${project.LIBRARY_VERSION}\"")
```
Comment (SeanoScribd) [inline Armadillo/build.gradle:34]: New gradle version doesn't automatically generate this for library projects

**Review by BarryFruitman — APPROVED** [review]

File: TestApp/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -25,7 +25,8 @@
         <meta-data android:name="com.google.android.gms.car.application"
             android:resource="@xml/automotive_app_desc"/>
 
-        <activity android:name=".presentation.MainActivity">
+        <activity android:name=".presentation.MainActivity"
+            android:exported="true">
```
Comment (rubeus90) [inline TestApp/src/main/AndroidManifest.xml:30]: Why do we need to mark these activities as exported ? We aren't invoking these activities from outside of the test app, are we ?

File: RELEASE.md [hunk 5]
```diff
@@ -1,5 +1,13 @@
 # Project Armadillo Release Notes
 
+## 1.0.4-SNAPSHOT
+- Upgraded Exoplayer version to 2.15.1
```
Comment (rubeus90) [inline RELEASE.md:4]: You updated this to `2.17.1` in the next commit

File: TestApp/src/main/AndroidManifest.xml (hunk 4 above)
Comment (SeanoScribd) [reply TestApp/src/main/AndroidManifest.xml:30]: Launch activities are technically always invoked from outside the app :)

`AudioPlayerActivity`  and `MediaButtonReceiver` get directly invoked from the notifications

File: RELEASE.md (hunk 5 above)
Comment (SeanoScribd) [reply RELEASE.md:4]: nice catch

**Review by rubeus90 — APPROVED** [review]

**Review by BarryFruitman — APPROVED** [review]

**Pull request merged**